### PR TITLE
Feature: add SetBufferSize() so that managed size can be dynamic

### PR DESCRIPTION
**Pull request by matthewvon**
This PR adds SetBufferSize() to the WriteBufferManager object.  This enables user code to adjust the global budget for write_buffers based upon other memory conditions such as growth in table reader memory as the dataset grows.

The buffer_size_ member variable is now atomic to match design of other changeable size_t members within WriteBufferManager.

This change is useful as is.  However, this change is also essential if someone decides they wanted to enable db_write_buffer_size modifications through the DB::SetOptions() API, i.e. no waste taking this as is.

Any format / spacing changes are due to clang-format as required by check-in automation.

### CODE REVIEW

**Comment by matthewvon** [comment]
@mrambacher This is the small PR I mention two weeks ago.  Finally got it prepared.

**Comment by matthewvon** [comment]
I looked at #7898.  My changes will fit there easily.  There are no logic changes related to buffer_size_ within that PR.  A manual edit will be required since PR 7898 moves some logic from .h to a new .cc file.  The comments make it unclear if #7898 is going to be accepted.  If it is, I am happy to update.

File: include/rocksdb/write_buffer_manager.h [hunk 1]
```diff
@@ -32,7 +32,9 @@ class WriteBufferManager {
 
   ~WriteBufferManager();
 
-  bool enabled() const { return buffer_size_ != 0; }
+  bool enabled() const {
+    return buffer_size_.load(std::memory_order_relaxed) != 0;
```
Comment (mrambacher) [inline include/rocksdb/write_buffer_manager.h:36]: return buffer_size() > 0

File: include/rocksdb/write_buffer_manager.h [hunk 2]
```diff
@@ -46,16 +48,19 @@ class WriteBufferManager {
...
 
   // Should only be called from write thread
   bool ShouldFlush() const {
     if (enabled()) {
       if (mutable_memtable_memory_usage() > mutable_limit_) {
         return true;
       }
-      if (memory_usage() >= buffer_size_ &&
-          mutable_memtable_memory_usage() >= buffer_size_ / 2) {
+      size_t local_size = buffer_size_.load(std::memory_order_relaxed);
```
Comment (mrambacher) [inline include/rocksdb/write_buffer_manager.h:61]: Use buffer_size()

File: include/rocksdb/write_buffer_manager.h [hunk 3]
```diff
@@ -46,16 +48,19 @@ class WriteBufferManager {
...
   }
-  size_t buffer_size() const { return buffer_size_; }
+  size_t buffer_size() const {
+    return buffer_size_.load(std::memory_order_relaxed);
+  }
 
   // Should only be called from write thread
   bool ShouldFlush() const {
     if (enabled()) {
       if (mutable_memtable_memory_usage() > mutable_limit_) {
```
Comment (mrambacher) [inline include/rocksdb/write_buffer_manager.h:58]: mutable_limit_ is based on the initial value of buffer_size.  Should it also be changed if the buffer size is changed?

**Review by mrambacher — COMMENTED** [review]
You need to write tests to show what the expected behavior is for various changes to the buffer_size value (enabled to disabled, disabled to enabled, size increased/decreased).

This is a public API change and will need to be documented in HISTORY.md

File: include/rocksdb/write_buffer_manager.h (hunk 1 above)
Comment (matthewvon) [reply include/rocksdb/write_buffer_manager.h:36]: done

File: include/rocksdb/write_buffer_manager.h (hunk 2 above)
Comment (matthewvon) [reply include/rocksdb/write_buffer_manager.h:61]: done

File: include/rocksdb/write_buffer_manager.h (hunk 3 above)
Comment (matthewvon) [reply include/rocksdb/write_buffer_manager.h:58]: unclear how I missed this.  thank you. done.

**Comment by matthewvon** [comment]
I now see ./java/rocksjni/write_buffer_manager.cc.  It is already a subset of the public API to this object.  Should it be updated as part of this PR?

File: HISTORY.md [hunk 4]
```diff
@@ -23,6 +23,7 @@
...
 * Add new Append and PositionedAppend APIs to FileSystem to bring the data verification information (data checksum information) from upper layer (e.g., WritableFileWriter) to the storage layer. In this way, the customized FileSystem is able to verify the correctness of data being written to the storage on time. Add checksum_handoff_file_types to DBOptions. User can use this option to control which file types (Currently supported file tyes: kWALFile, kTableFile, kDescriptorFile.) should use the new Append and PositionedAppend APIs to handoff the verification information. Currently, RocksDB only use crc32c to calculate the checksum for write handoff.
 * Add an option, `CompressionOptions::max_dict_buffer_bytes`, to limit the in-memory buffering for selecting samples for generating/training a dictionary. The limit is currently loosely adhered to.
+* Add new SetBufferSize API to WriteBufferManager to allow dynamic management of memory allotted to all write buffers.  This allows user code to adjust memory monitoring provided by WriteBufferManager as process memory needs change datasets grow and shrink.
```
Comment (akankshamahajan15) [inline HISTORY.md:26]: Can you add this change under the unreleased section at the top.

File: memtable/write_buffer_manager_test.cc [hunk 5]
```diff
@@ -47,7 +47,33 @@ TEST_F(WriteBufferManagerTest, ShouldFlush) {
   ASSERT_TRUE(wbf->ShouldFlush());
 
   wbf->FreeMem(7 * 1024 * 1024);
-  // 9MB total, 8MB mutable.
+  // 8MB total, 8MB mutable.
+  ASSERT_FALSE(wbf->ShouldFlush());
+
+  // change size: 8M limit, 7M mutable limit
```
Comment (akankshamahajan15) [inline memtable/write_buffer_manager_test.cc:53]: Can you add the test if buffer size is decreased as well.

**Review by akankshamahajan15 — APPROVED** [review]
Thanks for the change. LGTM. Added some minor comments.

File: HISTORY.md (hunk 4 above)
Comment (matthewvon) [reply HISTORY.md:26]: Done.

File: memtable/write_buffer_manager_test.cc (hunk 5 above)
Comment (matthewvon) [reply memtable/write_buffer_manager_test.cc:53]: @akankshamahajan15 Original buffer size was 10M and my added test lowers it to 8M.  I am confused as to what to write.

Comment (akankshamahajan15) [reply memtable/write_buffer_manager_test.cc:53]: My bad. Its fine.

**Comment by matthewvon** [comment]
@akankshamahajan15 Thank you.

**Comment by akankshamahajan15** [comment]
@matthewvon Can you update the branch. All internal tests passed and is ready to land.

**Comment by matthewvon** [comment]
@akankshamahajan15  Done.  Only HISTORY.md required manual merge.

**Pull request closed without merging**